File: client-lib-python/vaultcode/banner.py
"""Banner NON bloccante (INVARIANTE 1, art. 615-quinquies).

Emette un AVVISO quando un contesto non è verificato, ma lo script CONTINUA a
funzionare. Nessun blocco/uscita/eccezione propagata. Sink iniettabile (default:
logging su stderr; opzionale append su file). Parità con ``Banner.php``.
"""

from __future__ import annotations

import logging
import sys
from datetime import datetime, timezone

_log = logging.getLogger("vaultcode")


class Banner:
    def __init__(self, sink=None, log_file: str = ""):
        self._sink = sink
        self._log_file = log_file
        self._seen: set[str] = set()

    def _default_sink(self, msg: str) -> None:
        if self._log_file:
            try:
                with open(self._log_file, "a", encoding="utf-8") as fh:
                    fh.write(f"[{datetime.now(timezone.utc).isoformat()}] {msg}\n")
            except OSError:
                pass
        # sempre anche su logger/stderr, best-effort
        try:
            _log.warning(msg)
        except Exception:  # noqa: BLE001
            print(msg, file=sys.stderr)

    def notify(self, message: str) -> None:
        if message in self._seen:
            return
        self._seen.add(message)
        msg = f"[VaultCode] {message}"
        try:
            (self._sink or self._default_sink)(msg)
        except Exception:  # noqa: BLE001 - il banner non deve MAI rompere l'app
            pass
        # NB: nessun exit/raise. L'esecuzione prosegue.

File: lib/src/Banner.php
<?php
declare(strict_types=1);

namespace VaultCode;

/**
 * Banner NON bloccante (INVARIANTE 1, §6 art. 615-quinquies).
 *
 * Quando un contesto non è verificato (server irraggiungibile oltre il grace,
 * licenza non valida, ecc.) si emette un AVVISO, ma il software CONTINUA a
 * funzionare. Nessun blocco, nessuna cancellazione, nessun sabotaggio.
 *
 * Il sink è iniettabile (default: error_log) per integrarsi con il logging del
 * cliente senza interrompere l'esecuzione.
 */
final class Banner
{
    /** @var callable */
    private $sink;
    /** @var array<string,bool> dedup per messaggio nella stessa richiesta */
    private array $seen = [];

    public function __construct(?callable $sink = null)
    {
        $this->sink = $sink ?? static function (string $msg): void { error_log($msg); };
    }

    public function notify(string $message): void
    {
        if (isset($this->seen[$message])) { return; }
        $this->seen[$message] = true;
        ($this->sink)("[VaultCode] $message");
        // NB: nessun exit/die/throw. L'esecuzione prosegue.
    }
}
